src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractDocumentDiagnosticSourceProvider.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+internal abstract class AbstractDocumentDiagnosticSourceProvider(string name) : IDiagnosticSourceProvider
+{
+    public bool IsDocument => true;
+    public string Name => name;
+
+    public abstract ValueTask<ImmutableArray<IDiagnosticSource>> CreateDiagnosticSourcesAsync(RequestContext context, CancellationToken cancellationToken);
+
+    protected static TextDocument? GetOpenDocument(RequestContext context)
```

Comment:
I think it would be good if this implementation moved into AbstractDocumentPullDiagnosticHandler.GetOrderedDiagnosticSourcesAsync

then you don't have to call it in all the doc sources.